Sources/ForageSDK/Component/ForagePINTextField/Vault/RosettaPINTextField.swift

```diff
@@ -0,0 +1,180 @@
+        didSet {
+            if let widthConstraint = widthConstraint {
+                inputWidthConstraint?.constant = widthConstraint
+            }
+        }
+    }
+    
+    var heightConstraint: CGFloat? {
+        didSet {
+            if let heightConstraint = heightConstraint {
+                inputHeightConstraint?.constant = heightConstraint
+            }
+        }
+    }
+    
+    // MARK: - Initialization
+    
+    override init(frame: CGRect) {
+        textField = UITextField()
+        collector = CollectorFactory.createForage(environment: ForageSDK.shared.environment, textElement: textField)
+        super.init(frame: frame)
+        commonInit()
+    }
+    
+    required init?(coder aDecoder: NSCoder) {
+        textField = UITextField()
+        collector = CollectorFactory.createForage(environment: ForageSDK.shared.environment, textElement: textField)
+        super.init(coder: aDecoder)
+        commonInit()
+    }
+    
+    // MARK: - UITextFieldDelegate protocol methods
+    func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {
+        let currentValue = textField.text ?? ""
+        guard let valueRange = Range(range, in: currentValue) else { return false }
+        let newValue = currentValue.replacingCharacters(in: valueRange, with: string)
+        let isOnlyNumeric = newValue.allSatisfy{ $0.isNumber }
+        let isFourOrFewer = newValue.count <= 4
+        return isOnlyNumeric && isFourOrFewer
+    }
+    
+    // MARK: - Private API
+    
+    private func commonInit() {
+        addSubview(textField)
+        textField.delegate = self
+        textField.translatesAutoresizingMaskIntoConstraints = false
+        textField.keyboardType = UIKeyboardType.numberPad
+        textField.isSecureTextEntry = true
+        textField.textAlignment = .center
+        textField.anchor(
+            top: topAnchor,
+            leading: leadingAnchor,
+            bottom: bottomAnchor,
+            trailing: trailingAnchor,
+            centerXAnchor: nil,
+            padding: UIEdgeInsets(top: 0, left: 0, bottom: 0, right: 0)
+        )
+        inputWidthConstraint = textField.widthAnchor.constraint(greaterThanOrEqualToConstant: 342)
+        inputWidthConstraint?.isActive = true
+        inputHeightConstraint = textField.heightAnchor.constraint(greaterThanOrEqualToConstant: 36)
+        inputHeightConstraint?.isActive = true
+        textField.addTarget(self, action: #selector(textFieldDidChange), for: .editingChanged)
+    }
+    
+    @objc func textFieldDidChange(_ textField: UITextField) {
+        self._isEmpty = !textField.hasText
+        self._isValid = textField.text?.count == 4 && (textField.text?.allSatisfy { $0.isNumber } ?? false)
+        self._isComplete = self._isValid
+        self.delegate?.textFieldDidChange(self)
+    }
+    
+    func clearText() {
+        DispatchQueue.main.async {
+            self.textField.text = ""
```

Comment:
Does setting this value not trigger the `textFieldDidChange` event?

Reply:
Ah, it probably does. I had this code before I implemented the `textFieldDidChange` protocol method, so I didn't think about it. Will clean this up if that does work as expected 👍🏻

Reply:
[d300b26](https://github.com/teamforage/forage-ios-sdk/pull/169/commits/d300b2620dcc531cb22487bebfcfc60fc3d2f128)